### PR TITLE
fix(core): rank units by bias-corrected utility EMA in continual backprop (#2343)

### LINKED ISSUE #2343: CBP replaces units by the raw utility EMA, not the paper's bias-corrected utility (Dohare et al. 2024, Eq. 8) — it re-replaces its own freshly reset units
## Summary

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranks mature units by the **raw** utility EMA. The algorithm it cites ranks by the **bias-corrected** EMA.

Dohare, Hernandez-Garcia, Lan, Rahman, Mahmood & Sutton (2024), *"Loss of plasticity in deep continual learning"*, **Nature 632, 768–774**, Eq. 7–8:

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]        (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])               (8)
```

`a[l,i,t]` is the **per-unit age**. Algorithm 1 then selects the eligible units "with the smallest utility", i.e. smallest `u_hat`.

The authors' released implementation agrees — https://github.com/shibhansh/loss-of-plasticity, `lop/algos/gnt.py`:

```python
# GnT.update_utility
bias_correction = 1 - self.decay_rate ** self.ages[layer_idx]
...
self.bias_corrected_util[layer_idx] = self.util[layer_idx] / bias_correction

# GnT.test_features
new_features_to_replace = torch.topk(-self.bias_corrected_util[i][eligible_feature_indices],
                                     num_new_features_to_replace)[1]
```

`lop/algos/cbp_linear.py` and `lop/algos/cbp_conv.py` use the same per-unit-age correction.

## Why it does not cancel

The correction is per unit, and `_replace_one_unit` sets a replaced unit's `age` **and** `utility` back to 0. So eligible units in the same layer genuinely differ in age and therefore in warm-up bias. At the module defaults (`decay_rate=0.99`, `maturity_threshold=100`) a unit that has just become eligible reads at `1 - 0.99**100 = 0.6340` of its true utility, while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically targets the youngest eligible units — precisely the units the maturity threshold exists to protect. The paper states the motivation explicitly:

> "initializing the outgoing weight to zero makes the new unit vulnerable to immediate reinitialized as it has zero utility. To protect new units from a reinitialization for maturity threshold m number of updates."

The maturity threshold only defers the problem by `m` steps; without Eq. 8 the freshly reset unit is still the lowest-ranked unit the moment it matures.

## Falsifier (deterministic, pristine `main` @ `c9aba7b5`)

Both scenarios use `decay_rate=0.99`, `maturity_threshold=100`, and drive the real `update_utility` EMA.

**A — young-strong vs old-weak.** Unit 0 has seen a constant contribution of `1.0` for 100 steps; unit 1 a constant `0.9` for 700 steps.

```
raw EMA        : unit0=0.633967 (age 100)   unit1=0.899209 (age 700)
Eq. 8 corrected: unit0=1.000000             unit1=0.900001
true steady-state contributions: unit0=1.0  unit1=0.9

main            REPLACED UNIT: [0]   <- the strictly more useful unit
Eq. 8 / gnt.py  REPLACED UNIT: [1]
```

**B — repeat victim.** Four units with *identical* true contribution `1.0`; unit 2 was reset 100 steps ago (i.e. it is the unit CBP replaced last time and has only just matured).

```
raw EMA        : [0.999997, 0.999997, 0.633967, 0.999997]
Eq. 8 corrected: [0.999997, 0.999997, 1.000000, 0.999997]

main            REPLACED UNIT: [2]   <- re-replaces its own fresh unit
Eq. 8 / gnt.py  REPLACED UNIT: [0]
```

## Scope

- `core/continual_backprop.py` is **not** a registered evidence source (`alberta_framework/evaluation/*` registers `average_reward`, `baseline_optimizers`, `canonical_upgd`, `ftl_world_model`, `intelligence_amplification`, `interaction_features`, `normalizers`, `oak`, `options`, `optimizers`, `types`, `update_safety` — not this file). `alberta-evidence-status` reports byte-identical per-claim validity before and after.
- `benchmarks/ipmnist_screening.py` carries its own CBP implementation and does not import `core.continual_backprop`, so no screening shard, summary, or incumbent number moves.
- In-tree consumers are the public `alberta_framework` API (the robot track's `core/{...,continual_backprop,...}` import) and the tests.

## Precedent

Same defect class as #480 (`fix(features): bias-correct EMA utility outputs`, merged) and issue #431, which fixed the warm-up debias for `future_utility` / `feature_discovery`. `core/continual_backprop.py` was not covered by that change.

## Not changed here (separate, smaller)

Eligibility uses `age >= maturity_threshold`; the paper says "units with age **more than** m" and `gnt.py` uses `torch.where(self.ages > self.maturity_threshold)`. That is a one-step off-by-one and a separate variable; deliberately left out of the fix so this change has exactly one moving part.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2343.

## Summary
`_select_replacement_index` in `alberta_framework/core/continual_backprop.py` ranked mature units by their raw utility EMA `utility`. As described in Dohare et al. (2024, Eq. 8), units of varying ages have warm-up bias (`u_hat = u / (1 - eta ** age)`). Without Eq. 8 bias correction, freshly matured units (e.g. at age 100 with default decay 0.99) read at ~0.634 of their true steady-state contribution and were repeatedly re-replaced ahead of older, strictly less useful units.

## Proposed Changes
1. Updated `_select_replacement_index` to compute per-unit bias correction factor `1.0 - decay_rate ** age` and rank mature units by bias-corrected utility `utility / correction`.
2. Passed `config.decay_rate` from `replace_units_with_flags` into `_select_replacement_index`.
3. Added unit tests in `tests/test_continual_backprop.py` verifying that units with higher steady-state contributions but younger ages are properly protected over older, lower-utility units.

## Test Plan
- `uv run pytest tests/test_continual_backprop.py` (96 passed)
- `uv run ruff check alberta_framework/core/continual_backprop.py tests/test_continual_backprop.py` (clean)
- `uv run mypy alberta_framework/core/continual_backprop.py` (clean)

### CODE REVIEW

**Review by lumix5 — CHANGES_REQUESTED** [review]
Independent verification at the current head `dd2d082cb086a2c1819e06a112158b5e38c4ec2f` (the only prior review is an MLuber-bot approval, not a substantive independent current-head review; the #2569 twin was closed as superseded by this PR, so the burden lands here).

**Disposition: fix** — as written this is an unmeasured algorithm change to a core learner, not a defect repair.

- The module's own documented convention contradicts the change. `continual_backprop.py`'s module docstring ("Per-unit utility convention" + "Replacement" sections) specifies a plain EMA `u <- decay*u + (1-decay)*|a*g|` with selection of "the mature unit with the lowest utility in the layer", and states the gradient form "is what the original Dohare implementation uses in practice". This PR changes that selection behavior but leaves the module-level "Replacement" docstring stale — only the inner function docstring was updated, so the module now documents two different selection rules.
- The citation does not check out. The official implementation ([shibhansh/loss-of-plasticity](https://github.com/shibhansh/loss-of-plasticity), `lop/algos/cbp_linear.py`) tracks utility as a plain EMA (`features = features * decay_rate + (1 - decay_rate) * input`) and selects via `torch.topk(-self.util[eligible_feature_indices], k)` — smallest **raw** utility among mature units, with no `1 - eta**age` correction anywhere in tracking or selection. The maturity threshold is the paper's mechanism for protecting young units ([Dohare et al., Nature 2024](https://www.nature.com/articles/s41586-024-07711-7); e.g. maturity 1,000 in Fig. 1b). If "Eq. 8" really defines selection-time bias correction, quote it verbatim in the PR body — I could not find it in the paper or the reference code.
- The warm-up-bias phenomenon the PR describes is real (a freshly matured unit's EMA does read low at `1 - 0.99**100 ~ 0.63` of steady state), which is exactly why the paper gates eligibility by age. Deciding that the threshold is insufficient *for this repository's workloads* is a measured claim, not a citation — nothing here shows it.
- `alberta_framework.core.continual_backprop` is on the import surface consumed by the robot track, so the behavior change propagates beyond this repository's lanes unmeasured.
- The new test proves the arithmetic of the correction, not that corrected selection improves any measured result on this repository's lanes.

Path to acceptance (any one):
1. Pre-register a paired comparison on a CBP-consuming lane — plain-EMA selection baseline vs bias-corrected selection, shared seeds, one variable, win threshold stated before running — and post the result win or lose (a loss goes to `NEGATIVE_RESULTS_LEDGER.md`).
2. A maintainer explicitly blesses the fidelity change, the module-level "Replacement" docstring is updated to match, and the Eq. 8 citation is either quoted verbatim or corrected to describe this as an intentional deviation.

The implementation itself (guard on `correction <= 0`, passing `config.decay_rate`) is clean — the blocker is evidence, not code quality.

**Comment by lalalune** [comment]
Closing as superseded by merged #2345. The merged version applies per-unit EMA bias correction through the real replacement path and checks corrected utility eligibility. The replacement regression fails on main and the CBP/options panel passes all 206 tests.
